drivers/platform/xilinx/xilinx_spi_pl.c

```diff
@@ -0,0 +1,417 @@
+static inline uint32_t _rd_fifo(struct xspi_desc *xdesc, uint8_t *data,
+				uint32_t len)
+{
+	uint32_t rx_available;
+	uint32_t to_read;
+	uint32_t tmp;
+	uint32_t i;
+
+	rx_available = _read(xdesc, XSP_RFO_OFFSET) + 1;
+	to_read = min(rx_available, len);
+
+	for (i = 0; i < to_read; ++i) {
+		tmp = _read(xdesc, XSP_DRR_OFFSET);
+		if (data)
+			data[i] = tmp;
+	}
+
+	return to_read;
+}
+
+/* If start is 0 stop transfer, else start it */
+static void inline _xil_spi_start_transfer(struct spi_desc *desc, uint8_t start)
+{
+	struct xspi_desc	*xdesc;
+
+	xdesc = desc->extra;
+	if (start) {
+		_write(xdesc, XSP_SSR_OFFSET, ~(1 << desc->chip_select));
+		_update(xdesc, XSP_CR_OFFSET, 0, XSP_CR_TRANS_INHIBIT_MASK);
+	} else {
+		_update(xdesc, XSP_CR_OFFSET, XSP_CR_TRANS_INHIBIT_MASK,
+			XSP_CR_TRANS_INHIBIT_MASK);
+		_write(xdesc, XSP_SSR_OFFSET, xdesc->disable_slaves_mask);
+	}
+	xdesc->cs_asserted = start;
+}
+
+
+/* SPI polled transfer */
+static int32_t xil_spi_write_and_read_pl(struct spi_desc *desc, uint8_t *data,
+		uint16_t bytes_number)
+{
+	struct xspi_desc	*xdesc;
+	uint32_t		tx;
+	uint32_t		rx;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	_update_mode(desc);
+
+	rx = 0;
+	tx = 0;
+	_xil_spi_start_transfer(desc, 1);
+	while (rx < bytes_number) {
+		if (tx < bytes_number)
+			tx += _wr_fifo(xdesc, data + tx, bytes_number - tx);
+		rx += _rd_fifo(xdesc, data + rx, bytes_number - rx);
+	}
+	_xil_spi_start_transfer(desc, 0);
+
+	return SUCCESS;
+}
+
+/* Read/write FIFOs to/from buffers and update it structure accordingly.
+ * Return */
+static inline uint8_t _transfer_and_update(struct xspi_desc *xdesc,
+		struct msg_iterator *it)
+{
+	uint32_t		i;
+	uint32_t		total;
+	uint32_t		len;
+	uint8_t			cs_change;
+
+	total = 0;
+	cs_change = 0;
+	do {
+		len = it->msg->bytes_number - it->i;
+		if (it->is_write)
+			i = _wr_fifo(xdesc, it->buff + it->i, len);
+		else if (it->buff)
+			i = _rd_fifo(xdesc, it->buff + it->i, len);
+		else
+			/* Upper layer don't provided a read buffer */
+			i = _rd_fifo(xdesc, NULL, len);
+
+		total += i;
+		it->i += i;
+		/* If current msg have been sent/read go to next one */
+		if (it->i == it->msg->bytes_number) {
+			cs_change = it->msg->cs_change;
+			it->msg = it->msg->next;
+			it->i = 0;
+			if (it->is_write)
+				it->buff = it->msg->tx_buff;
+			else
+				it->buff = it->msg->rx_buff;
+			if (cs_change)
+				return 1;
+		}
+	} while (total < xdesc->fifo_depth && it->msg);
+	/* Until FIFO is full or empty or no more data to transfer */
+
+	return 0;
+}
+
+/* SPI polled transfer of multiple messages */
+static int32_t xil_spi_transfer_pl(struct spi_desc *desc,
+				   struct spi_msg_list *head)
+{
+	struct xspi_desc	*xdesc;
+	uint8_t			cs_change;
+	struct msg_iterator	rx;
+	struct msg_iterator	tx;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	_update_mode(desc);
+
+	rx.msg = head;
```

Comment:
this iterator just complicates things. Just use msg_list you already have there the tx and rx buffer an you actually don't use the index outside of _transfer_and_update so in can be a tmp variable there.

Reply:
I will think of something to make the function cleaner